File: RobotLibrary/multipleDevices/multi.py
########################################################################
# Copyright 2020 Roku, Inc.
#
#Licensed under the Apache License, Version 2.0 (the "License");
#you may not use this file except in compliance with the License.
#You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#Unless required by applicable law or agreed to in writing, software
#distributed under the License is distributed on an "AS IS" BASIS,
#WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#See the License for the specific language governing permissions and
#limitations under the License.
########################################################################

import sys
import asyncio
import subprocess
import json
from string import Template
import codecs
if sys.platform == "win32":
    asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())

async def read_stream(stream, cb):
    while True:
        line = await stream.readline()
        if line:
            cb(line)
        else:
            break


async def stream_subprocess(cmd, stdout_cb, stderr_cb):
    try:
        process = await asyncio.create_subprocess_exec(
            *cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
        )

        await asyncio.wait(
            [
                read_stream(process.stdout, stdout_cb),
                read_stream(process.stderr, stderr_cb),
            ]
        )
        rc = await process.wait()
        return process.pid, rc
    except OSError as e:
        return e

async def prc(*aws):
    await  asyncio.gather(*aws)

def execute(*aws):
    asyncio.run(prc(*aws))

def printer(label):
    def pr(*args):
        print(label, *args)

    return pr

def runners(cmds):
    for cmd in cmds:
        out = printer(f"{cmd['name']}.stdout")
        err = printer(f"{cmd['name']}.stderr")
        yield stream_subprocess(cmd['command'], out, err)

def getCmdList(config):
    devices = config['devices']
    outputdir = config['outputdir']
    if devices == None:
        raise Exception("devices field is empty")
    
    for device in devices:
        cmd = {
            'command': [sys.executable, "-m", "robot.run"],
            'name': device
        }
        deviceFields = devices[device]
        for key in deviceFields:
            cmd['command'].extend(['--variable', f'{key}:{deviceFields[key]}'])
        cmd['command'].extend(['--outputdir', f'{outputdir}/{device}'])
        cmd['command'].append(config['test'])
        yield cmd

def createGeneralFile(config, fileName, title):
    devicesList = []
    outputdir = config['outputdir']
    for device in config['devices']:
        devicesList.append({"name": device, "src": f'{device}/{fileName}.html'})
    with  codecs.open("multipleDevices/template.html", 'r', 'utf-8') as template:
        res = Template(template.read()).safe_substitute(content=devicesList, title=title)
    with open(f"{outputdir}/multi_device_{fileName}.html", "w") as new_file:  
        new_file.write(res)

if __name__ == "__main__":
    try:
        config_path = sys.argv[1]
        with open(config_path) as myfile:
            data = json.load(myfile)
        if data['server_path'] == None:
            raise Exception("server_path field is required")
        else:
            process = subprocess.Popen(data['server_path'])
        cmdList = getCmdList(data)
        execute(*runners(cmdList))
        process.kill()
        createGeneralFile(data, "log", "Logs")
        createGeneralFile(data, "report", "Reports")
    except IndexError:
        print("Config path is required as first argument")
    except FileNotFoundError:
        print('This file doesn\'t exist')
    except Exception as e:
        print(e)

File: jsLibrary/multipleDevices/multi.js
///////////////////////////////////////////////////////////////////////////
// Copyright 2020 Roku, Inc.
//
//Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
//////////////////////////////////////////////////////////////////////////

const { spawn } = require('child_process');
const fs = require("fs");

function getCmdList(config) {
    const devices = config.devices;
    if (devices === undefined) {
        throw new Error("devices field is empty");
    }
    let cmds = [];
    for(const device in devices) {
        let deviceFields = devices[device];
        let cmd = {
            options: '',
            name: device 
        }
        for(const field in deviceFields) {
            cmd.options += `--${field}=${deviceFields[field]} `;
        }
        cmds.push(cmd);
    }
    return cmds;
}

function startTest(cmds, config) {
    let numberOfProcess = cmds.length;
    const {test, outputdir, server_path} = config;
    const webDriver = spawn(server_path);
    const isWin = process.platform === "win32";
    const executeCmd = isWin ? 'mocha.cmd' : 'mocha';
    cmds.forEach(({options, name}) => {
        let child = spawn(executeCmd, [`${test}`,'--reporter', 'mochawesome', '--reporter-options', `reportDir=${outputdir},reportFilename=${name}`, options]);
        child.on('error', (messasge) => {
            console.log(`${name} error: ${messasge}`);
        });
        child.stdout.on('data', (data) => {
            console.log(`${name} stdout:\n${data}`);
        });

        child.on('exit', () => {
            numberOfProcess--;
            if (numberOfProcess === 0) {
                webDriver.kill();
                createGeneralReport(config);
            }
        });
    });
}

function createGeneralReport(config) {
    let deviceList = [];
    const {devices, outputdir} = config;
    for(const device in devices) {
        deviceList.push({'name': device, 'src': `${device}.html`});
    }
    let template = fs.readFileSync('multipleDevices/template.html', 'utf8');
    template = template.replace('$content', JSON.stringify(deviceList));
    template = template.replace('$title', 'Report');
    fs.writeFile(`${outputdir}/multi_device_report.html`, template, (err) => {
        if (err) throw err;
        console.log(`Multi-device report saved to ${outputdir}/multi_device_report.html`);
    });
}

function checkRequiredFields(json) {
    const fields = ['devices', 'server_path', 'outputdir', 'test'];
    fields.forEach( field => {
       if (json[field] === undefined) {
          throw new Error(`Please, add field \'${field}\' to config`);
       }
    });
}

try {
    const args = process.argv;
    const configPath = args[2];
    if (configPath === undefined) {
        throw new Error('Path to config is required');
    }
    const config = fs.readFileSync(configPath);
    const jsonData = JSON.parse(config);
    checkRequiredFields(jsonData);
    const cmdList = getCmdList(jsonData);
    startTest(cmdList, jsonData);
} catch (e) {
    console.log(e);
}